src/test/java/org/padaiyal/utilities/aayvalar/regex/RegexUtilityTest.java

```diff
@@ -0,0 +1,544 @@
+      "\\w+, Hello World, 1, 'Hello'",
+      "\\d+, 23, 2, '23'",
+      "\\w*, Hello. World stuff, 4, 'Hello,,,World'",
+  })
+  void testGetFirstNumbersMatchesWithoutCapturingGroups(
+      String regexPattern,
+      String input,
+      int topNumbers,
+      @ConvertWith(StringArrayConverter.class) String[] expectedMatches
+  ) throws InvalidRegexException {
+    List<String> expectedTopNumberMatches = Arrays.asList(expectedMatches);
+
+    List<String> actualTopNumberMatches =
+        RegexUtility.getFirstNumbersMatches(regexPattern, 0, topNumbers, input)
+            .parallelStream()
+            .map((matches) -> {
+              Assertions.assertEquals(1, matches.size());
+              return matches.get(0);
+            })
+            .collect(Collectors.toList());
+    Assertions.assertEquals(expectedTopNumberMatches, actualTopNumberMatches);
+  }
+
+  /**
+   * Test getting the top first numbers of matches with capturing groups in the regex expression.
+   *
+   * @param regexPattern The regex pattern to test.
+   * @param input        The input string.
+   * @param topNumbers   The top first numbers of matches to retrieve.
+   * @throws InvalidRegexException If an invalid regex exception is provided.
+   */
+  @ParameterizedTest
+  @CsvSource({
+      "([A-Za-z0-9][A-Za-z0-9-]+)\\.([A-Za-z0-9][A-Za-z0-9-]+), "
+          + "some.domain hello.com, 3",
+      "(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\"
+          + ".(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\"
+          + ".(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?)\\"
+          + ".(25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9]?), 192.125.1.1 127.0.0 125.1.1.1, 2"
+  })
+  void testGetFirstNumbersMatchesWithCapturingGroups(
+      String regexPattern,
+      String input,
+      int topNumbers
+  ) throws InvalidRegexException {
+
+    List<String> regexPatternAndGroups = getCaptureGroupsRegex(regexPattern);
+    regexPatternAndGroups.add(0, regexPattern);
+    List<List<String>> actualMatches =
+        RegexUtility.getFirstNumbersMatches(regexPattern, 0, topNumbers, input);
+
+    Assertions.assertTrue(topNumbers >= actualMatches.size());
+
+    actualMatches.parallelStream()
+        .forEach(matches ->
+            IntStream.range(0, matches.size())
+                .forEach(index -> {
+                  String match = matches.get(index);
+                  String regexExpression = regexPatternAndGroups.get(index);
+
+                  Assertions.assertTrue(
+                      RegexUtility.matches(regexExpression, 0, match)
+                  );
+                }));
+  }
+
+  @ParameterizedTest
+  @CsvSource({
+      "\\w",
+      "\\w+",
+      "\\w*",
+      "\\s",
+      "\\s+",
+      "\\s*",
+      "\\d",
+      "\\d+",
+      "\\d*",
+      "[abz]",
+      "[abz]*",
+      "[abz]+",
+      "hello.*",
+      ".*",
+      "hello.+",
+      ".+",
+      "hello.",
+      "."
+  })
+  void testFillRandomValues(String regexToTest) throws InvalidRegexException {
```

Comment:
Try combinations of different regex expressions.

Reply:
I have added more tests scenarios. Let me know if more are required.